homeassistant/components/tplink/entity.py

```diff
@@ -275,55 +301,44 @@ def _entities_for_device_and_its_children[_E: CoordinatedTPLinkEntity](
     *,
     feature_type: Feature.Type,
     entity_class: type[_E],
+    descriptions: tuple[TPLinkFeatureEntityDescription, ...] | None = None,
+    new_description_generator: Callable[[Feature], TPLinkFeatureEntityDescription]
+    | None = None,
+    child_coordinators: list[TPLinkDataUpdateCoordinator] | None = None,
 ) -> list[_E]:
     """Create entities for device and its children.
 
     This is a helper that calls *_entities_for_device* for the device and its children.
     """
     entities: list[_E] = []
+    # Add parent entities before children so via_device id works.
+    entities.extend(
+        _entities_for_device(
+            device,
+            coordinator=coordinator,
+            feature_type=feature_type,
+            entity_class=entity_class,
+            descriptions=descriptions,
+            new_description_generator=new_description_generator,
+        )
+    )
     if device.children:
         _LOGGER.debug("Initializing device with %s children", len(device.children))
-        for child in device.children:
+        for idx, child in enumerate(device.children):
+            if child_coordinators:
+                child_coordinator = child_coordinators[idx]
+            else:
+                child_coordinator = coordinator
             entities.extend(
                 _entities_for_device(
                     child,
-                    coordinator=coordinator,
+                    coordinator=child_coordinator,
                     feature_type=feature_type,
                     entity_class=entity_class,
                     parent=device,
+                    descriptions=descriptions,
+                    new_description_generator=new_description_generator,
                 )
             )
 
-    entities.extend(
-        _entities_for_device(
-            device,
-            coordinator=coordinator,
-            feature_type=feature_type,
-            entity_class=entity_class,
-        )
-    )
-
     return entities
-
-
-def _description_for_feature[_D: EntityDescription](
```

Comment:
Dropped from here as the typing doesn't really work and it seems better to let the sensor and switch implement their own methods